### ISSUE TITLE
Running tsc manually crashes with "Parameter symbol already has a cached type which differs from newly assigned type"

### ISSUE THREAD

**Issue by sirbrillig** [Automattic/wp-calypso#66894]
### Quick summary

You can manually run the TypeScript type checks for the client directory with the command `yarn tsc -p client --noEmit`. This is helpful for finding TS errors before making a commit. 

However, now when I run that command, I get the following unhelpful error:

```
/Users/payton/Code/wp-calypso/node_modules/typescript/lib/tsc.js:96888
                throw e;
                ^

Error: Debug Failure. Expected [object Object] === [object Object]. Parameter symbol already has a cached type which differs from newly assigned type
    at assignParameterType (/Users/payton/Code/wp-calypso/node_modules/typescript/lib/tsc.js:64862:26)
    at assignContextualParameterTypes (/Users/payton/Code/wp-calypso/node_modules/typescript/lib/tsc.js:64827:21)
    at contextuallyCheckFunctionExpressionOrObjectLiteralMethod (/Users/payton/Code/wp-calypso/node_modules/typescript/lib/tsc.js:65229:29)
    at checkFunctionExpressionOrObjectLiteralMethod (/Users/payton/Code/wp-calypso/node_modules/typescript/lib/tsc.js:65208:13)
    at checkExpressionWorker (/Users/payton/Code/wp-calypso/node_modules/typescript/lib/tsc.js:66805:28)
    at checkExpression (/Users/payton/Code/wp-calypso/node_modules/typescript/lib/tsc.js:66707:38)
    at checkExpressionWithContextualType (/Users/payton/Code/wp-calypso/node_modules/typescript/lib/tsc.js:66396:28)
    at inferTypeArguments (/Users/payton/Code/wp-calypso/node_modules/typescript/lib/tsc.js:62906:39)
    at chooseOverload (/Users/payton/Code/wp-calypso/node_modules/typescript/lib/tsc.js:63566:53)
    at resolveCall (/Users/payton/Code/wp-calypso/node_modules/typescript/lib/tsc.js:63403:26)
```

This same error happens if I run `yarn tsc --build client/tsconfig.json` which appears to run fine in the CI environment. I tried starting with a fresh `node_modules` and removing `.tsc-cache` but neither has any effect.

It seems to have been introduced by https://github.com/Automattic/wp-calypso/pull/66817. Before that commit, running the command works as expected. Afterward, it produces the error above. 

However, I can't see how that PR could have caused any problems. I tried manually undoing the changes it made and the error continues.

### Steps to reproduce

1. `git checkout f0d3672658` (this is the merge commit which matches up to https://github.com/Automattic/wp-calypso/pull/66817)
2. `yarn tsc -p client --noEmit`
3. Notice that it displays a debug error.
4. `git checkout f0d3672658~1` (this is the commit before https://github.com/Automattic/wp-calypso/pull/66817)
5. `yarn tsc -p client --noEmit`
6. Notice that the process completes successfully.

### What you expected to happen

`tsc` should run and display any TypeScript errors.

### What actually happened

`tsc` fails.

**Comment by sirbrillig** [comment]
🤔 Hang on...

I ran `tsc` inside the `client` directory, which _does_ appear to work.

```
cd client && tsc
```
... and now running `yarn tsc -p client --noEmit` from the top-level directory seems to work again 🙃 

Ok. Nevermind, I guess?

**Comment by noahtallen** [comment]
Super weird! Maybe a random bug with how yarn resolves binaries?

**Comment by sirbrillig** [comment]
🤷 Well, hopefully if anyone else encounters this, they'll find my "solution" helpful.

**Comment by sirbrillig** [comment]
Reopening this because it really happens a lot. I'd love to find a real solution.

**Comment by danielbachhuber** [comment]
This issue happens to me every time I miss a TypeScript error.

For example, the following diff produces a crash:

```diff
diff --git a/client/my-sites/hosting/sftp-card/ssh-keys.tsx b/client/my-sites/hosting/sftp-card/ssh-keys.tsx
index 7298936fa7..512416acab 100644
--- a/client/my-sites/hosting/sftp-card/ssh-keys.tsx
+++ b/client/my-sites/hosting/sftp-card/ssh-keys.tsx
@@ -18,7 +18,7 @@ import { useAttachSshKeyMutation } from './use-attach-ssh-key';
 import './ssh-keys.scss';
 
 interface SshKeysProps {
-	siteId: number;
+	sited: number;
 	siteSlug: string;
 	username: string;
 	disabled: boolean;
```

<img width="835" alt="image" src="https://user-images.githubusercontent.com/36432/222296112-624ff094-bc6d-4b9f-bfb0-93c59da83174.png">

<img width="1547" alt="image" src="https://user-images.githubusercontent.com/36432/222296304-47cbaa11-58db-4f1f-9862-bd5c324def2f.png">

Maybe this is a known issue though? https://github.com/microsoft/TypeScript/issues/50773

@Automattic/team-calypso-frameworks Is TypeScript supposed to crash this hard when there's an error? I would've expected more friendly error output...

**Comment by tyxla** [comment]
I hit this same error when trying to run `yarn tsc --noEmit --project client/tsconfig.json` so there must be something going on. I'm investigating.

**Comment by tyxla** [comment]
~Got a fix for this #73991 that resolved it for me locally~
~Working on a fix, this still needs some work.~
Finished the fix for this #73991 that resolved it for me locally - cc @danielbachhuber and @sirbrillig.
